tests/subsys/audio_module/src/functional_test.c

```diff
@@ -0,0 +1,1801 @@
+
+#include "fakes.h"
+#include "audio_module/audio_module.h"
+#include "common.h"
+
+K_THREAD_STACK_DEFINE(mod_stack, TEST_MOD_THREAD_STACK_SIZE);
+K_MEM_SLAB_DEFINE(mod_data_slab, TEST_MOD_DATA_SIZE, FAKE_FIFO_MSG_QUEUE_SIZE, 4);
+
+static struct mod_context mod_context, test_mod_context;
+static struct mod_config mod_config = {
+	.test_int1 = 5, .test_int2 = 4, .test_int3 = 3, .test_int4 = 2};
+static struct audio_module_functions mod_functions_null = {.open = NULL,
+							   .close = NULL,
+							   .configuration_set = NULL,
+							   .configuration_get = NULL,
+							   .start = NULL,
+							   .stop = NULL,
+							   .data_process = NULL};
+static struct audio_module_functions mod_functions_populated = {
+	.open = test_open_function,
+	.close = test_close_function,
+	.configuration_set = test_config_set_function,
+	.configuration_get = test_config_get_function,
+	.start = test_stop_start_function,
+	.stop = test_stop_start_function,
+	.data_process = test_data_process_function};
+static char *test_base_name = "Test base name";
+static struct audio_module_description mod_description = {.name = "Test base name",
+							  .type = AUDIO_MODULE_TYPE_IN_OUT,
+							  .functions = &mod_functions_null};
+static struct audio_module_description test_from_description, test_to_description;
+static struct audio_module_parameters mod_parameters = {
+	.description = &mod_description,
+	.thread = {.stack = mod_stack,
+		   .stack_size = TEST_MOD_THREAD_STACK_SIZE,
+		   .priority = TEST_MOD_THREAD_PRIORITY,
+		   .data_slab = &mod_data_slab,
+		   .data_size = TEST_MOD_DATA_SIZE}};
+struct audio_module_parameters test_mod_parameters;
+static struct audio_module_handle handle;
+static struct mod_context *handle_context;
+static struct data_fifo mod_fifo_tx, mod_fifo_rx;
+
+/**
+ * @brief Set the minimum for a handle.
+ *
+ * @param handle[in/out]       The handle to the module instance
+ * @param state[in]            Pointer to the module's state
+ * @param name[in]             Pointer to the module's base name
+ * @param description[in/out]  Pointer to the modules description
+ */
+static void test_handle_set(struct audio_module_handle *hdl, enum audio_module_state state,
+			    char *name, struct audio_module_description *description)
+{
+	hdl->state = state;
+	description->name = name;
+	hdl->description = description;
+}
+
+/**
+ * @brief Simple test thread with handle NULL.
+ *
+ * @param handle[in]  The handle to the module instance
+ *
+ * @return 0 if successful, error value
+ */
+static int test_thread_handle(struct audio_module_handle const *const handle)
+{
+	/* Execute thread */
+	while (1) {
+		zassert_not_null(handle, NULL, "handle is NULL!");
+
+		k_sleep(K_MSEC(100));
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Simple function to start a test thread with handle.
+ *
+ * @param handle[in/out]  The handle to the module instance
+ *
+ * @return 0 if successful, error value
+ */
+static int start_thread(struct audio_module_handle *handle)
```

Comment:
Nitpick: Usually we have the verb last:
```suggestion
static int thread_start(struct audio_module_handle *handle)
```